File: src/main/java/org/jcodec/api/transcode/PixelStoreImpl.java
package org.jcodec.api.transcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import org.jcodec.common.model.ColorSpace;
import org.jcodec.common.model.Picture;

public class PixelStoreImpl implements PixelStore {
    private Random rnd = new Random();
    private int bound;
    private List<Picture> buffers;

    public PixelStoreImpl() {
        buffers = new ArrayList<Picture>();
        bound = 10;
    }

    public int getBound() {
        return bound;
    }

    public void setBound(int bound) {
        this.bound = bound;
    }

    @Override
    public LoanerPicture getPicture(int width, int height, ColorSpace color) {
        for (Picture picture : buffers) {
            if (picture.getWidth() == width && picture.getHeight() == height && picture.getColor() == color) {
                buffers.remove(picture);
                return new LoanerPicture(picture, 1);
            }
        }
        return new LoanerPicture(Picture.create(width, height, color), 1);
    }

    @Override
    public void putBack(LoanerPicture frame) {
        frame.decRefCnt();
        if (frame.unused()) {
            Picture pixels = frame.getPicture();
            pixels.setCrop(null);
            if (buffers.size() == bound) {
                buffers.set(rnd.nextInt(bound), pixels);
            } else {
                buffers.add(pixels);
            }
        }
    }

    @Override
    public void retake(LoanerPicture frame) {
        frame.incRefCnt();
    }
}

File: src/main/java/org/jcodec/api/transcode/PixelStoreImpl.kt
package org.jcodec.api.transcode

import org.jcodec.api.transcode.PixelStore.LoanerPicture
import org.jcodec.common.model.ColorSpace
import org.jcodec.common.model.Picture
import java.util.*

class PixelStoreImpl : PixelStore {
    private val buffers: MutableList<Picture>
    override fun getPicture(width: Int, height: Int, color: ColorSpace): LoanerPicture {
        for (picture in buffers) {
            if (picture.width == width && picture.height == height && picture.color == color) {
                buffers.remove(picture)
                return LoanerPicture(picture, 1)
            }
        }
        return LoanerPicture(Picture.create(width, height, color), 1)
    }

    override fun putBack(frame: LoanerPicture) {
        frame.decRefCnt()
        if (frame.unused()) {
            val pixels = frame.picture
            pixels.crop = null
            buffers.add(pixels)
        }
    }

    override fun retake(frame: LoanerPicture) {
        frame.incRefCnt()
    }

    init {
        buffers = ArrayList()
    }
}